File: solutions/go/collatz-conjecture/2/collatz_conjecture.go
package collatzconjecture
import "errors"

func CollatzConjecture(n int) (steps int, err error) {
    if n<=0 {
        return 0, errors.New("input must be a positive integer")
    }
    
	for steps =0 ;n>1; steps++ {
        if n%2==0 {
            n /= 2
        } else  {
			n = 3*n + 1            
        }
    }
    return steps, nil
}

File: solutions/ocaml/collatz-conjecture/2/collatz_conjecture.ml
open Base

let collatz_conjecture num = 
  let rec aux num steps =
    match num with
      | 1 -> Ok steps
      | n when n % 2 = 0 -> aux (n/2) (steps+1)
      | _ -> aux (3*num+1) (steps+1)
  in 
  if num<=0  then 
       Error "Only positive integers are allowed"
  else aux num 0
